src/libos/crates/async-rt/src/executor.rs

```diff
@@ -125,40 +152,220 @@ impl Executor {
     }
 
     fn pick_thread_for(&self, task: &Arc<Task>) -> usize {
+        let last_thread_id = task.sched_info().last_thread_id() as usize;
+        let priority = task.sched_info().priority();
+        let candidates_num = core::cmp::max((self.parallelism / 2) as usize, 1);
+        let latency_weights: Vec<f64> = {
+            let tmp_latencies: Vec<u64> = (0..self.parallelism as usize)
+                .map(|idx| self.latencies[idx].load(Ordering::Relaxed))
+                .collect();
+            let max_latency = *tmp_latencies.iter().max().unwrap();
+            let min_latency = *tmp_latencies.iter().min().unwrap();
+            if max_latency == min_latency {
+                (0..self.parallelism as usize).map(|_| 0.0).collect()
+            } else {
+                tmp_latencies
+                    .iter()
+                    .map(|l| 1.0 - (l - min_latency) as f64 / (max_latency - min_latency) as f64)
+                    .collect()
+            }
+        };
+        let length_weights: Vec<f64> = {
+            let worker_lens: Vec<usize> = (0..self.parallelism as usize)
+                .map(|idx| self.workers[idx].relax_len(*priority))
+                .collect();
+            let max_len = *worker_lens.iter().max().unwrap();
+            let min_len = *worker_lens.iter().min().unwrap();
+            if max_len == min_len {
+                (0..self.parallelism as usize).map(|_| 0.0).collect()
+            } else {
+                worker_lens
+                    .iter()
+                    .map(|l| 1.0 - (l - min_len) as f64 / (max_len - min_len) as f64)
+                    .collect()
+            }
+        };
+
         let affinity = task.sched_info().affinity().read();
         assert!(!affinity.is_empty());
-        let mut thread_id = task.sched_info().last_thread_id() as usize;
-        while !affinity.get(thread_id) {
-            thread_id = (thread_id + 1) % Affinity::max_threads();
-        }
+        let thread_id = affinity
+            .iter()
+            .enumerate()
+            .chain(affinity.iter().enumerate())
+            .skip(last_thread_id)
+            .filter(|(_idx, bit)| *bit)
+            .take(candidates_num)
+            .map(move |(idx, _bit)| {
+                let len = self.workers[idx].relax_len(*priority);
+                if len == Worker::QUEUED_TASKS_MAX_SIZE {
+                    return (idx, 0);
+                }
+                let affinity_weight = (idx == last_thread_id) as u64 as f64; // 0.0 or 1.0
+                let weight = length_weights[idx] * 700.0
+                    + latency_weights[idx] * 200.0
+                    + affinity_weight * 100.0;
+                (idx, weight as u64)
+            })
+            .max_by_key(|x| x.1)
+            .unwrap()
+            .0;
         drop(affinity);
 
         task.sched_info().set_last_thread_id(thread_id as u32);
         thread_id
     }
 
+    fn try_rebalance_workload(&self) {
```

Comment:
Is is possible to encapsulate the rebalancing logic? This file is simply too large.